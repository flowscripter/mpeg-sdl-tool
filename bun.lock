{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "@flowscripter/mpeg-sdl-tool",
      "dependencies": {
        "@flowscripter/dynamic-cli-framework": "latest",
        "@flowscripter/mpeg-sdl-parser": "latest",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "latest",
      },
    },
  },
  "packages": {
    "@flowscripter/dynamic-cli-framework": ["@flowscripter/dynamic-cli-framework@1.1.4", "", { "dependencies": { "emphasize": "^7.0.0", "fastest-levenshtein": "^1.0.16", "figlet": "^1.8.0", "highlight.js": "^11.11.1", "prettier": "^3.5.3", "supports-color": "^10.0.0", "ts-log": "^2.2.7" }, "peerDependencies": { "typescript": "^5.8.3" } }, "sha512-gCPxAW97BnpN+8Ohq1v19POJm0fG6si5uNu68BpEKTnOsEL7Lp18Y1WyUNx+I1z+K8ya8mTptxPcSwtAdOHkWg=="],

    "@flowscripter/mpeg-sdl-parser": ["@flowscripter/mpeg-sdl-parser@1.1.9", "", { "dependencies": { "typescript-parsec": "^0.3.4" }, "peerDependencies": { "typescript": "^5.8.3" } }, "sha512-H/vOowsUq+LYP1VDWnV1xtkSZxysk8XSkcqPdS9aoySKKIT1j7eZkfvHdBNGM9r1jczroZt+bW2Rbsxw+nHEXw=="],

    "@types/bun": ["@types/bun@1.2.9", "", { "dependencies": { "bun-types": "1.2.9" } }, "sha512-epShhLGQYc4Bv/aceHbmBhOz1XgUnuTZgcxjxk+WXwNyDXavv5QHD1QEFV0FwbTSQtNq6g4ZcV6y0vZakTjswg=="],

    "@types/hast": ["@types/hast@3.0.4", "", { "dependencies": { "@types/unist": "*" } }, "sha512-WPs+bbQw5aCj+x6laNGWLH3wviHtoCv/P3+otBhbOhJgG8qtpdAMlTCxLtsTWA7LH1Oh/bFCHsBn0TPS5m30EQ=="],

    "@types/node": ["@types/node@22.13.10", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-I6LPUvlRH+O6VRUqYOcMudhaIdUVWfsjnZavnsraHvpBwaEyMN29ry+0UVJhImYL16xsscu0aske3yA+uPOWfw=="],

    "@types/unist": ["@types/unist@3.0.3", "", {}, "sha512-ko/gIFJRv177XgZsZcBwnqJN5x/Gien8qNOn0D5bQU/zAzVf9Zt3BlcUiLqhV9y4ARk0GbT3tnUiPNgnTXzc/Q=="],

    "@types/ws": ["@types/ws@8.5.14", "", { "dependencies": { "@types/node": "*" } }, "sha512-bd/YFLW+URhBzMXurx7lWByOu+xzU9+kb3RboOteXYDfW+tr+JZa99OyNmPINEGB/ahzKrEuc8rcv4gnpJmxTw=="],

    "bun-types": ["bun-types@1.2.9", "", { "dependencies": { "@types/node": "*", "@types/ws": "*" } }, "sha512-dk/kOEfQbajENN/D6FyiSgOKEuUi9PWfqKQJEgwKrCMWbjS/S6tEXp178mWvWAcUSYm9ArDlWHZKO3T/4cLXiw=="],

    "chalk": ["chalk@5.4.1", "", {}, "sha512-zgVZuo2WcZgfUEmsn6eO3kINexW8RAE4maiQ8QNs8CtpPCSyMiYsULR3HQYkm3w8FIA3SberyMJMSldGsW+U3w=="],

    "dequal": ["dequal@2.0.3", "", {}, "sha512-0je+qPKHEMohvfRTCEo3CrPG6cAzAYgmzKyxRiYSSDkS6eGJdyVJm7WaYA5ECaAD9wLB2T4EEeymA5aFVcYXCA=="],

    "devlop": ["devlop@1.1.0", "", { "dependencies": { "dequal": "^2.0.0" } }, "sha512-RWmIqhcFf1lRYBvNmr7qTNuyCt/7/ns2jbpp1+PalgE/rDQcBT0fioSMUpJ93irlUhC5hrg4cYqe6U+0ImW0rA=="],

    "emphasize": ["emphasize@7.0.0", "", { "dependencies": { "@types/hast": "^3.0.0", "chalk": "^5.0.0", "highlight.js": "~11.9.0", "lowlight": "~3.1.0" } }, "sha512-jdFCDyt+YetBXO12VwK4AiLsMCvkZ3IBxMVIJddB+25EwIL0VETBgpvPkJl63+JyAgaQ5Wja10qWMoXXC95JNg=="],

    "fastest-levenshtein": ["fastest-levenshtein@1.0.16", "", {}, "sha512-eRnCtTTtGZFpQCwhJiUOuxPQWRXVKYDn0b2PeHfXL6/Zi53SLAzAHfVhVWK2AryC/WH05kGfxhFIPvTF0SXQzg=="],

    "figlet": ["figlet@1.8.0", "", { "bin": { "figlet": "bin/index.js" } }, "sha512-chzvGjd+Sp7KUvPHZv6EXV5Ir3Q7kYNpCr4aHrRW79qFtTefmQZNny+W1pW9kf5zeE6dikku2W50W/wAH2xWgw=="],

    "highlight.js": ["highlight.js@11.11.1", "", {}, "sha512-Xwwo44whKBVCYoliBQwaPvtd/2tYFkRQtXDWj1nackaV2JPXx3L0+Jvd8/qCJ2p+ML0/XVkJ2q+Mr+UVdpJK5w=="],

    "lowlight": ["lowlight@3.1.0", "", { "dependencies": { "@types/hast": "^3.0.0", "devlop": "^1.0.0", "highlight.js": "~11.9.0" } }, "sha512-CEbNVoSikAxwDMDPjXlqlFYiZLkDJHwyGu/MfOsJnF3d7f3tds5J3z8s/l9TMXhzfsJCCJEAsD78842mwmg0PQ=="],

    "prettier": ["prettier@3.5.3", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-QQtaxnoDJeAkDvDKWCLiwIXkTgRhwYDEQCghU9Z6q03iyek/rxRh/2lC3HB7P8sWT2xC/y5JDctPLBIGzHKbhw=="],

    "supports-color": ["supports-color@10.0.0", "", {}, "sha512-HRVVSbCCMbj7/kdWF9Q+bbckjBHLtHMEoJWlkmYzzdwhYMkjkOwubLM6t7NbWKjgKamGDrWL1++KrjUO1t9oAQ=="],

    "ts-log": ["ts-log@2.2.7", "", {}, "sha512-320x5Ggei84AxzlXp91QkIGSw5wgaLT6GeAH0KsqDmRZdVWW2OiSeVvElVoatk3f7nicwXlElXsoFkARiGE2yg=="],

    "typescript": ["typescript@5.8.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-p1diW6TqL9L07nNxvRMM7hMMw4c5XOo/1ibL4aAIGmSAt9slTE1Xgw5KWuof2uTOvCg9BY7ZRi+GaF+7sfgPeQ=="],

    "typescript-parsec": ["typescript-parsec@0.3.4", "", {}, "sha512-6RD4xOxp26BTZLopNbqT2iErqNhQZZWb5m5F07/UwGhldGvOAKOl41pZ3fxsFp04bNL+PbgMjNfb6IvJAC/uYQ=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],

    "emphasize/highlight.js": ["highlight.js@11.9.0", "", {}, "sha512-fJ7cW7fQGCYAkgv4CPfwFHrfd/cLS4Hau96JuJ+ZTOWhjnhoeN1ub1tFmALm/+lW5z4WCAuAV9bm05AP0mS6Gw=="],

    "lowlight/highlight.js": ["highlight.js@11.9.0", "", {}, "sha512-fJ7cW7fQGCYAkgv4CPfwFHrfd/cLS4Hau96JuJ+ZTOWhjnhoeN1ub1tFmALm/+lW5z4WCAuAV9bm05AP0mS6Gw=="],
  }
}
