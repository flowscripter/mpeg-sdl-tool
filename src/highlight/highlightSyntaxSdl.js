/**
 * Highlight.js SDL syntax highlighting definition
 */
export default function (hljs) {
  const COMMENTS = hljs.COMMENT("//", "$");

  const BUILT_INS = [
    "lengthof",
  ];

  const KEYWORDS = [
    "abstract",
    "aligned",
    "break",
    "case",
    "class",
    "computed",
    "const",
    "default",
    "do",
    "else",
    "expandable",
    "extends",
    "for",
    "if",
    "legacy",
    "map",
    "reserved",
    "switch",
    "while",
  ];

  const TYPES = [
    "unsigned",
    "bit",
    "int",
    "base64string",
    "float",
    "utf16string",
    "utf8string",
    "utf8list",
    "utfstring",
  ];

  const NUMBERS = {
    className: "number",
    variants: [
      {
        // Literal binary
        begin: "(0b[0-1]{1,4})",
      },
      {
        // Literal hexadecimal
        begin: "(0x[0-9a-fA-F]{1,4})",
      },
      {
        // Literal integer, decimal, float
        begin: "\\b\\d+(\\.\\d+)?(E\\d+)?",
      },
    ],
  };

  const MULTIPLE_CHARACTER_LITERALS = {
    className: "string",
    begin: /'[^']*'/,
  };

  const STRING_LITERALS = {
    className: "string",
    begin: /u?"[^"]*"/,
  };

  const PUNCTUATORS = {
    className: "punctuation",
    match: /[\(\)\[\]{};:]/,
  };

  const OPERATORS = {
    className: "operator",
    match: /[\.+\-*/%<>=!&\|]/,
  };

  return {
    name: "SDL",
    disableAutodetect: true,
    keywords: {
      keyword: KEYWORDS,
      type: TYPES,
    },
    built_ins: BUILT_INS,
    contains: [
      MULTIPLE_CHARACTER_LITERALS,
      STRING_LITERALS,
      COMMENTS,
      NUMBERS,
      PUNCTUATORS,
      OPERATORS,
    ],
  };
}
