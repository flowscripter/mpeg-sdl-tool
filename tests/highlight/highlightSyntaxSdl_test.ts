import fs from "node:fs/promises";
import process from "node:process";
import path from "node:path";
import { describe, expect, test } from "bun:test";
import {
  DefaultPrinterService,
  DefaultSyntaxHighlighterService,
} from "@flowscripter/dynamic-cli-framework";
import highlightSyntaxSdl from "../../src/highlight/highlightSyntaxSdl";
import TtyTerminal from "@flowscripter/dynamic-cli-framework/src/service/printer/terminal/TtyTerminal";
import { Writable } from "node:stream";
import TtyStyler from "@flowscripter/dynamic-cli-framework/src/service/printer/terminal/TtyStyler";
import {
  createLenientSdlParser,
  prettyPrint,
  SdlStringInput,
} from "@mpeggroup/mpeg-sdl-parser";
import { buildAst } from "@mpeggroup/mpeg-sdl-parser/src/ast/buildAst";

describe("Highlight Plugin SDL tests", () => {
  test("Highlight test", async () => {
    const printer = new DefaultPrinterService(
      Writable.toWeb(process.stdout),
      Writable.toWeb(process.stderr),
      true,
      true,
      new TtyTerminal(process.stderr),
      new TtyStyler(3),
    );
    const syntaxHighlighterService = new DefaultSyntaxHighlighterService();

    printer.colorEnabled = true;

    syntaxHighlighterService.registerSyntax("sdl", highlightSyntaxSdl);
    syntaxHighlighterService.colorFunction = printer.color.bind(printer);

    const sampleSdlSpecification = await fs.readFile(
      path.join(__dirname, "../sample_specifications/various_elements.sdl"),
    ).then((buffer) => buffer.toString());

    const sdlStringInput = new SdlStringInput(sampleSdlSpecification);
    const parser = await createLenientSdlParser();

    const sdlParseTree = parser.parse(sdlStringInput);
    const sdlSpecification = buildAst(sdlParseTree, sdlStringInput);

    const prettifiedSdlSpecification = await prettyPrint(
      sdlSpecification,
      sdlStringInput,
    );
    const colorScheme = {
      keyword: "#f00000",
      string: "#f0f000",
      number: "#00f000",
      punctuation: "#00f0f0",
      operator: "#0000f0",
    };

    const highlighted = syntaxHighlighterService.highlight(
      prettifiedSdlSpecification,
      "sdl",
      colorScheme,
    );

    expect(highlighted).toMatchSnapshot();
  });
});
