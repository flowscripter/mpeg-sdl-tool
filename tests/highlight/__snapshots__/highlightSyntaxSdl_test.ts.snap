// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlight Plugin SDL tests Highlight test 1`] = `
"// repeated packets in bitstream
//
// should be declared in a wrapping Bitstream class
\x1B[38;2;240;0;0mclass\x1B[39m transport_packet \x1B[38;2;0;240;240m{\x1B[39m
  unsigned int\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m8\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m sync_byte \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m0x47\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  bit\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m1\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m transport_error_indicator\x1B[38;2;0;240;240m;\x1B[39m
  bit\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m1\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m payload_unit_start_indicator\x1B[38;2;0;240;240m;\x1B[39m
  bit\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m1\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m transport_priority\x1B[38;2;0;240;240m;\x1B[39m
  unsigned int\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m13\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m // odd but valid comment position
   PID\x1B[38;2;0;240;240m;\x1B[39m
  bit\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m2\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m\x1B[38;2;0;0;240m*\x1B[39m transport_scrambling_control\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0maligned\x1B[39m\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m16\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m bit\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m2\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m adaptation_field_control \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m0x0000\x1B[39m\x1B[38;2;0;0;240m.\x1B[39m\x1B[38;2;0;240;0m0000\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mcomputed\x1B[39m unsigned int N \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m184\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mcomputed\x1B[39m float test \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m1\x1B[39me1\x1B[38;2;0;240;240m;\x1B[39m
  utf8string foo \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;240;240;0mu"bar"\x1B[39m \x1B[38;2;240;240;0mu"foo"\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  base64string foo2 \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;240;240;0m"BAR"\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mcomputed\x1B[39m int dummy1 \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;240;240;0m'ABCD'\x1B[39m \x1B[38;2;240;240;0m'01AB'\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mcomputed\x1B[39m int dummy2 \x1B[38;2;0;0;240m=\x1B[39m lengthof\x1B[38;2;0;240;240m(\x1B[39mtransport_scrambling_control\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;0;240m+\x1B[39m N\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mif\x1B[39m \x1B[38;2;0;240;240m(\x1B[39madaptation_field_control \x1B[38;2;0;0;240m=\x1B[39m\x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m0b10\x1B[39m \x1B[38;2;0;0;240m|\x1B[39m\x1B[38;2;0;0;240m|\x1B[39m adaptation_field_control \x1B[38;2;0;0;240m=\x1B[39m\x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m0b11\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
    adaptation_field data\x1B[38;2;0;240;240m;\x1B[39m // previously declared class

    // N is 184 minus the number of bytes in the adaptation_field
    N \x1B[38;2;0;0;240m=\x1B[39m N \x1B[38;2;0;0;240m-\x1B[39m // and another
    \x1B[38;2;0;240;0m1\x1B[39m // and another odd comment
     \x1B[38;2;0;0;240m-\x1B[39m data\x1B[38;2;0;0;240m.\x1B[39madaptation_field_length\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;0;240;240m}\x1B[39m
  \x1B[38;2;240;0;0mif\x1B[39m \x1B[38;2;0;240;240m(\x1B[39madaptation_field_control \x1B[38;2;0;0;240m=\x1B[39m\x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m0b01\x1B[39m \x1B[38;2;0;0;240m|\x1B[39m\x1B[38;2;0;0;240m|\x1B[39m adaptation_field_control \x1B[38;2;0;0;240m=\x1B[39m\x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m0b11\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
    bit\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m8\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m data_byte\x1B[38;2;0;240;240m[\x1B[39mN\x1B[38;2;0;240;240m]\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;0;240;240m}\x1B[39m
\x1B[38;2;0;240;240m}\x1B[39m

\x1B[38;2;240;0;0mmap\x1B[39m Offsets \x1B[38;2;0;240;240m(\x1B[39mint\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
  \x1B[38;2;0;240;0m0b00\x1B[39m, \x1B[38;2;0;240;240m{\x1B[39m \x1B[38;2;0;240;0m1024\x1B[39m \x1B[38;2;0;240;240m}\x1B[39m,
  \x1B[38;2;0;240;0m0b01\x1B[39m, \x1B[38;2;0;240;240m{\x1B[39m \x1B[38;2;0;240;0m2048\x1B[39m \x1B[38;2;0;240;240m}\x1B[39m
\x1B[38;2;0;240;240m}\x1B[39m

\x1B[38;2;240;0;0mexpandable\x1B[39m\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m20\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;240;0;0mclass\x1B[39m odd\x1B[38;2;0;240;240m(\x1B[39mint a, B b\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;240;0;0mextends\x1B[39m odder\x1B[38;2;0;240;240m(\x1B[39ma\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m:\x1B[39m bit\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m2\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m id \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m2\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
  \x1B[38;2;240;0;0mreserved\x1B[39m unsigned int\x1B[38;2;0;240;240m(\x1B[39mOffsets\x1B[38;2;0;240;240m)\x1B[39m offset\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mcomputed\x1B[39m int myArray\x1B[38;2;0;240;240m[\x1B[39m\x1B[38;2;0;240;0m2\x1B[39m\x1B[38;2;0;240;240m]\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mcomputed\x1B[39m int i \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m1\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  Strange s\x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m1\x1B[39m, \x1B[38;2;0;240;0m2\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mwhile\x1B[39m \x1B[38;2;0;240;240m(\x1B[39mi \x1B[38;2;0;0;240m<\x1B[39m \x1B[38;2;0;240;0m2\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
    i\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;0;240;240m}\x1B[39m
  i \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m2\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mdo\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
    i\x1B[38;2;0;0;240m-\x1B[39m\x1B[38;2;0;0;240m-\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;0;240;240m}\x1B[39m \x1B[38;2;240;0;0mwhile\x1B[39m \x1B[38;2;0;240;240m(\x1B[39mi \x1B[38;2;0;0;240m>\x1B[39m \x1B[38;2;0;240;0m0\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mif\x1B[39m \x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m1\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
    i\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;0;240;240m}\x1B[39m
  \x1B[38;2;240;0;0melse\x1B[39m \x1B[38;2;240;0;0mif\x1B[39m \x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m1\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
    i\x1B[38;2;0;0;240m-\x1B[39m\x1B[38;2;0;0;240m-\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;0;240;240m}\x1B[39m
  \x1B[38;2;240;0;0melse\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
    i\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;0;240;240m}\x1B[39m
  \x1B[38;2;240;0;0mif\x1B[39m \x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m1\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m
    \x1B[38;2;240;0;0mif\x1B[39m \x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;0m2\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m
      i\x1B[38;2;0;0;240m-\x1B[39m\x1B[38;2;0;0;240m-\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
    \x1B[38;2;240;0;0melse\x1B[39m
      i\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mswitch\x1B[39m \x1B[38;2;0;240;240m(\x1B[39mi\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
    \x1B[38;2;240;0;0mcase\x1B[39m \x1B[38;2;0;240;0m0\x1B[39m\x1B[38;2;0;240;240m:\x1B[39m
    \x1B[38;2;240;0;0mcase\x1B[39m \x1B[38;2;0;240;0m1\x1B[39m\x1B[38;2;0;240;240m:\x1B[39m
      // nothing
      \x1B[38;2;240;0;0mbreak\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
    \x1B[38;2;240;0;0mdefault\x1B[39m\x1B[38;2;0;240;240m:\x1B[39m
      i\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
    // foo
  \x1B[38;2;0;240;240m}\x1B[39m
\x1B[38;2;0;240;240m}\x1B[39m

\x1B[38;2;240;0;0mclass\x1B[39m odd2\x1B[38;2;0;240;240m(\x1B[39mint a\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
  \x1B[38;2;240;0;0mcomputed\x1B[39m int x \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m0\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;240;0;0mfor\x1B[39m \x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;240;0;0mcomputed\x1B[39m int j \x1B[38;2;0;0;240m=\x1B[39m \x1B[38;2;0;240;0m0\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m j \x1B[38;2;0;0;240m<\x1B[39m \x1B[38;2;0;240;0m2\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m j\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
    x\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;0;240m+\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m
  \x1B[38;2;0;240;240m}\x1B[39m
  \x1B[38;2;240;0;0mfor\x1B[39m \x1B[38;2;0;240;240m(\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m j \x1B[38;2;0;0;240m>\x1B[39m \x1B[38;2;0;240;0m0\x1B[39m\x1B[38;2;0;240;240m;\x1B[39m j\x1B[38;2;0;0;240m-\x1B[39m\x1B[38;2;0;0;240m-\x1B[39m\x1B[38;2;0;240;240m)\x1B[39m \x1B[38;2;0;240;240m{\x1B[39m
  \x1B[38;2;0;240;240m}\x1B[39m
\x1B[38;2;0;240;240m}\x1B[39m // until EOF
"
`;
